### PR TITLE
VASP Optics Masking, and plotting

### PR DESCRIPTION
**Pull request by jmmshn**
## Vasp optics masking and plotting features

Summary:

- Allow masking of matrix elements without recreating the object
- Allow users to access a filtered version of the matrix element data for analysis

You can now pass a `mask` kwarg in the calculation and get the portion of the dielectric function contributed by the transitions that you selected.

### CODE REVIEW

**Comment by jmmshn** [comment]
@janosh, ready to merge this is failing due to something unrelated.

**Comment by janosh** [comment]
> @janosh, ready to merge this is failing due to something unrelated.

Not sure where the error on master crept in. I was just investigating and working on a fix in #2695. It seems to be related to the pinned version of `spglib`:

https://github.com/materialsproject/pymatgen/blob/e01d044986d6b321ee0254a6c9c89a5edd693935/requirements.txt#L10

File: pymatgen/io/vasp/optics.py [hunk 1]
```diff
@@ -171,6 +176,69 @@ def _use_default(param, default):
...
+
+        x_val = []
+        y_val = []
+        text = []
+        (
+            nb,
+            _,
+            nk,
+            nspin,
+        ) = cderm.shape[:4]
```
Comment (janosh) [inline pymatgen/io/vasp/optics.py:221]: This could prob fit on 1 line and be more readable?

File: pymatgen/io/vasp/optics.py [hunk 2]
```diff
@@ -301,14 +373,43 @@ def epsilon_imag(
...
+            return egrid, np.zeros_like(egrid, dtype=np.complex_)
+        raise e
+
+    (
+        _,
+        _,
+        nk,
+        nspin,
+    ) = cder.shape[:4]
+    iter_idx = [
```
Comment (janosh) [inline pymatgen/io/vasp/optics.py:398]: Same here.

File: pymatgen/io/vasp/optics.py [hunk 3]
```diff
@@ -171,6 +176,69 @@ def _use_default(param, default):
             eps += 1.0 + 0.0j
         return egrid, eps
 
+    def plot_me_data(self, idir: int, jdir, mask: npt.NDArray | None = None, min_val: float = 0.0):
```
[the remark this answers is not in the document]
Comment (jmmshn) [reply pymatgen/io/vasp/optics.py]: changed to `plot_weighted_transition_data`.

File: pymatgen/io/vasp/optics.py (hunk 1 above)
Comment (jmmshn) [reply pymatgen/io/vasp/optics.py:227]: `black` decided to do this.

File: pymatgen/io/vasp/optics.py (hunk 2 above)
Comment (jmmshn) [reply pymatgen/io/vasp/optics.py:404]: Same here.

File: pymatgen/io/vasp/optics.py [hunk 4]
```diff
@@ -110,7 +112,8 @@ def from_directory(cls, directory: Path | str):
         """Construct a DielectricFunction from a directory containing vasprun.xml and WAVEDER files."""
         d_ = Path(directory)
         vrun = Vasprun(d_ / "vasprun.xml")
-        waveder = Waveder.from_binary(d_ / "WAVEDER")
+        is_gamma = "gamma" in vrun.generator["subversion"]
```
Comment (jmmshn) [inline pymatgen/io/vasp/optics.py]: This is kinda hacky but the "gamma-only" is defined in `src/version.F` so I think we are OK.

File: pymatgen/io/vasp/optics.py (hunk 1 above)
Comment (janosh) [reply pymatgen/io/vasp/optics.py:224]: `black` will leave the following as is:

```py
nb, _, nk, nspin = cderm.shape[:4]
```

Comment (jmmshn) [reply pymatgen/io/vasp/optics.py]: Oh wow... the comma!

**Comment by jmmshn** [comment]
OK, so due to the fact that VASP can be compiled with different dtypes I'm just gonna set `data_type="complex64"` and let the users pick when they are running for the `Waverder` object, `gamma_only` is just not enough information.

In the optics module, I'm trying to read the larger dtypes first to force a value error during reshaping. then trying smaller dtypes.

**Comment by janosh** [comment]
Thanks @jmmshn! 👍

**Pull request merged**